test/test_HeadingController.py

```diff
@@ -0,0 +1,354 @@
+
+        # No switch
+        self.assertFalse(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+        # Adjust parameters
+        mock_speed *= 2
+        mock_time += 0.5 * sailbot_constants.SWITCH_INTERVAL
+        mock_heading_error = 1.5 * sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        # No switch
+        self.assertFalse(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+    # Exit TACKABLE due to slow speed
+
+    def test_slowSpeed_exitTackable(self):
+        mock_speed = sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS + 0.1
+        mock_time = 0
+
+        # Enter TACKABLE from UNKNOWN
+        hc = HeadingController(mock_speed, mock_time, ControlModes.UNKNOWN.value)
+
+        # Make speed low and adjust time to meet the switch interval
+        mock_speed -= 0.2
+        mock_time += sailbot_constants.SWITCH_INTERVAL
+
+        self.assertTrue(hc.switchControlMode(
+            heading_error=0.5,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+
+        # Should not be in TACKABLE or UNKNOWN
+        self.assertTrue(hc.getControlModeID() != ControlModes.TACKABLE.value)
+        self.assertTrue(hc.getControlModeID() != ControlModes.UNKNOWN.value)
+
+    # Exit TACKABLE due to reaching the heading
+
+    def test_reachedHeading_exitTackable(self):
+        mock_speed = sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS + 0.1
+        mock_time = 0
+
+        # Enter TACKABLE from UNKNOWN
+        hc = HeadingController(mock_speed, mock_time, ControlModes.UNKNOWN.value)
+
+        # Adjust time to meet the switch interval
+        mock_time += sailbot_constants.SWITCH_INTERVAL
+        mock_heading_error = -0.5 * sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        # A switch occurs
+        self.assertTrue(hc.switchControlMode(
+            heading_error=mock_heading_error,
+            boat_speed=mock_speed,
+            current_time=mock_time
+        ))
+
+        # Should not have changed modes
+        self.assertEqual(hc.getControlModeID(), ControlModes.TACKABLE.value)
+
+    # Exit TACKABLE due to timeout
+    # This test assumes that the timeout interval is larger than the switch interval
+
+    def test_timeout_exitTackable(self):
+        mock_speed = sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS + 0.1
+        mock_time = 0
+
+        # Enter TACKABLE from UNKNOWN
+        hc = HeadingController(mock_speed, mock_time, ControlModes.UNKNOWN.value)
+
+        # Adjust time to meet the switch interval
+        mock_time += sailbot_constants.SWITCH_INTERVAL
+
+        # A switch should NOT occur
```

Comment:
just to confirm what is the difference between switch interval and timeout interval? I actually thought we want to switch in this case haha

Reply:
This the opposite to one of the comments that you made earlier. Even if we meet the switch interval and we attempt to switch, if we do not meet the exit condition then we do not switch. The documentation describes when True is returned and False is returned. Hopefully that will answer your question.

Reply:
you're right, thanks for clarifying